source/opt/decoration_manager.cpp

```diff
@@ -18,14 +18,91 @@
 #include <set>
 #include <stack>
 
+#include "ir_context.h"
+
 namespace spvtools {
 namespace opt {
 namespace analysis {
 
-void DecorationManager::RemoveDecorationsFrom(uint32_t id) {
-  auto const ids_iter = id_to_decoration_insts_.find(id);
+void DecorationManager::RemoveDecorationsFrom(
+    uint32_t id, std::function<bool(const ir::Instruction&)> pred) {
+  const auto ids_iter = id_to_decoration_insts_.find(id);
   if (ids_iter == id_to_decoration_insts_.end()) return;
-  id_to_decoration_insts_.erase(ids_iter);
+
+  TargetData& decorations_info = ids_iter->second;
+  auto context = module_->context();
+  std::vector<ir::Instruction*> insts_to_kill;
+  const bool is_group = !decorations_info.decorate_insts.empty();
+
+  // Schedule all direct decorations for removal if instructed as such by
+  // |pred|.
+  for (ir::Instruction* inst : decorations_info.direct_decorations)
+    if (pred(*inst)) insts_to_kill.push_back(inst);
+
+  // For all groups being directly applied to |id|, remove |id| (and the
+  // literal if |inst| is an OpGroupMemberDecorate) from the instruction
+  // applying the group.
+  std::unordered_set<const ir::Instruction*> indirect_decorations_to_remove;
+  for (ir::Instruction* inst : decorations_info.indirect_decorations) {
+    assert(inst->opcode() == SpvOpGroupDecorate ||
+           inst->opcode() == SpvOpGroupMemberDecorate);
+    const uint32_t stride = inst->opcode() == SpvOpGroupDecorate ? 1u : 2u;
+    bool was_modified = false;
+    for (uint32_t i = 1u; i < inst->NumInOperands();) {
+      if (inst->GetSingleWordInOperand(i) != id) {
+        i += stride;
+        continue;
+      }
+
+      const uint32_t last_operand_index = inst->NumInOperands() - stride;
+      if (i < last_operand_index)
+        inst->GetInOperand(i) = inst->GetInOperand(last_operand_index);
+      // Remove the associated literal, if it exists.
+      if (stride == 2u) {
+        if (i < last_operand_index)
+          inst->GetInOperand(i + 1u) =
+              inst->GetInOperand(last_operand_index + 1u);
+        inst->RemoveInOperand(last_operand_index + 1u);
```

Comment:
The use if `pred` seem inconsistent.  If the decoration is applied through a group you do not check it.  That does not seems right.

Reply:
You are right, I completely missed that case.